hare/algorithm.go

```diff
@@ -0,0 +1,230 @@
+	pubKey      crypto.PublicKey
+	layerId     LayerId
+	oracle      Rolacle // roles oracle
+	signing     Signing
+	network     NetworkService
+	startTime   time.Time // TODO: needed?
+	inbox       chan *pb.HareMessage
+	knowledge   []*pb.HareMessage
+	stateMutex  sync.Mutex
+	roundMsg    []byte
+	// TODO: add knowledge of notify messages. What is the life-cycle of such messages? persistent according to Julian
+}
+
+func NewConsensusProcess(key crypto.PublicKey, layer LayerId, s Set, oracle Rolacle, signing Signing, p2p NetworkService) *ConsensusProcess {
+	proc := &ConsensusProcess{}
+	proc.State = State{0, 0, s, Certificate{}}
+	proc.Closer = NewCloser()
+	proc.pubKey = key
+	proc.layerId = layer
+	proc.oracle = oracle
+	proc.signing = signing
+	proc.network = p2p
+	proc.knowledge = make([]*pb.HareMessage, 0, InitialKnowledgeSize)
+	m, err := proc.buildStatusMessage()
+	if err != nil {
+		log.Error("could not build status message")
+		proc.roundMsg = nil
+	} else {
+		proc.roundMsg = m
+	}
+
+	return proc
+}
+
+func (proc *ConsensusProcess) Start() error {
+	if !proc.startTime.IsZero() { // called twice on same instance
+		log.Error("ConsensusProcess has already been started.")
+		return StartInstanceError(errors.New("instance already started"))
+	}
+
+	proc.startTime = time.Now()
+
+	go proc.eventLoop()
+
+	return nil
+}
+
+func (proc *ConsensusProcess) createInbox(size uint32) chan *pb.HareMessage {
+	proc.inbox = make(chan *pb.HareMessage, size)
+	return proc.inbox
+}
+
+func (proc *ConsensusProcess) eventLoop() {
+	log.Info("Start listening")
+
+	ticker := time.NewTicker(RoundDuration)
+
+	for {
+		select {
+		case msg := <-proc.inbox: // msg event
+			proc.handleMessage(msg)
+		case <-ticker.C: // next round event
+			proc.nextRound()
+		case <-proc.CloseChannel(): // close event
+			log.Info("Stop event loop, instance aborted")
+			return
+		}
+	}
+}
+
+func (proc *ConsensusProcess) handleMessage(m *pb.HareMessage) {
+	// Note: layer is already verified by the broker
+
+	// verify iteration
+	if m.Message.K != proc.k {
+		return
+	}
+
+	// validate signature
+	data, err := proto.Marshal(m.Message)
+	if err != nil {
+		log.Error("Failed marshaling inner message")
+		return
+	}
+	if !proc.signing.Validate(data, m.InnerSig) {
+		log.Warning("invalid message signature detected")
+		return
+	}
+
+	// validate role
+	if !proc.oracle.ValidateRole(roleFromIteration(m.Message.K),
+		RoleRequest{m.PubKey, LayerId{NewBytes32(m.Message.Layer)}, m.Message.K},
+		Signature(m.Message.RoleProof)) {
+		log.Warning("invalid role detected")
+		return
+	}
+
+	// continue process msg by round
+	proc.processMessageByRound(m)
+
+	// TODO: decide how we store messages maybe we don't need raw msgs
+	proc.knowledge = append(proc.knowledge, m)
+}
+
+func (proc *ConsensusProcess) nextRound() {
+	log.Info("End of round: %d", proc.k)
+
+	// send message if available
+	if proc.roundMsg != nil {
+		proc.network.Broadcast(ProtoName, proc.roundMsg)
```

Comment:
Broadcast can fail and in such case it returns an error, check for it

Reply:
Fixed